Source/DafnyCore/Resolver/Resolver.cs

```diff
@@ -2670,270 +2675,104 @@ public class ModuleBindings {
-          // NewTypeDecl's and SubsetTypeDecl's were already processed in the loop above, except for any witness clauses
-          var dd = (RedirectingTypeDecl)d;
-          if (dd.Witness != null) {
-            var prevErrCnt = reporter.Count(ErrorLevel.Error);
-            var codeContext = new CodeContextWrapper(dd, dd.WitnessKind == SubsetTypeDecl.WKind.Ghost);
-            scope.PushMarker();
-            if (d is not TopLevelDeclWithMembers topLevelDecl || !topLevelDecl.AcceptThis) {
-              scope.AllowInstance = false;
-            }
-            ResolveExpression(dd.Witness, new ResolutionContext(codeContext, false));
-            scope.PopMarker();
-            ConstrainSubtypeRelation(dd.Var.Type, dd.Witness.Type, dd.Witness, "witness expression must have type '{0}' (got '{1}')", dd.Var.Type, dd.Witness.Type);
-            SolveAllTypeConstraints();
-            if (reporter.Count(ErrorLevel.Error) == prevErrCnt) {
-              CheckTypeInference(dd.Witness, dd);
-            }
-            if (reporter.Count(ErrorLevel.Error) == prevErrCnt && dd.WitnessKind == SubsetTypeDecl.WKind.Compiled) {
-              ExpressionTester.CheckIsCompilable(this, dd.Witness, codeContext);
-            }
-          }
-          if (d is TopLevelDeclWithMembers dm) {
-            ResolveClassMemberBodies(dm);
-          }
-        } else {
-          if (!(d is IteratorDecl)) {
-            // Note, attributes of iterators are resolved by ResolvedIterator, after registering any names in the iterator signature
-            ResolveAttributes(d, new ResolutionContext(new NoContext(d.EnclosingModuleDefinition), false));
-          }
-          if (d is IteratorDecl) {
-            var iter = (IteratorDecl)d;
-            ResolveIterator(iter);
-            ResolveClassMemberBodies(iter);  // resolve the automatically generated members
-          } else if (d is DatatypeDecl) {
-            var dt = (DatatypeDecl)d;
-            foreach (var ctor in dt.Ctors) {
-              ResolveAttributes(ctor, new ResolutionContext(new NoContext(d.EnclosingModuleDefinition), false));
-            }
-            // resolve any default parameters
-            foreach (var ctor in dt.Ctors) {
-              scope.PushMarker();
-              scope.AllowInstance = false;
-              ctor.Formals.ForEach(p => scope.Push(p.Name, p));
-              ResolveParameterDefaultValues(ctor.Formals, ResolutionContext.FromCodeContext(dt));
-              scope.PopMarker();
-            }
-            // resolve members
-            ResolveClassMemberBodies(dt);
-          } else if (d is TopLevelDeclWithMembers) {
-            var dd = (TopLevelDeclWithMembers)d;
-            ResolveClassMemberBodies(dd);
-          }
-        }
-        allTypeParameters.PopMarker();
+        var boundsDiscoveryVisitor = new BoundsDiscoveryVisitor(reporter);
+        boundsDiscoveryVisitor.VisitDeclarations(declarations);
       }
 
+      // Build call graph.
```

Comment:
Comment seems redundant based on the code `CallGraphBuilder.Build(declarations, reporter);`